### PR TITLE
Pattern modifications

### PR DESCRIPTION
**Pull request by codeFather2**

### CODE REVIEW

File: Sources/PT.PM.Matching.Tests/PatternMatchingTests.cs [hunk 1]
```diff
@@ -110,6 +110,25 @@ public void Match_PasswordCheckInsideStatement(string patternData)
...
+                Languages = new HashSet<Language> { Php.Language },
+                Node = new PatternAny("password")
+            };
+            patternsRepository.Add(patternsConverter.ConvertBack(new List<PatternRoot>() { pattern }));
+
+            WorkflowResult workflowResult = workflow.Process();
+            IEnumerable<MatchResultDto> matchResults = workflowResult.MatchResults.ToDto();
+            patternsRepository.Clear();
+
+            int expectedCount = 3;
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching.Tests/PatternMatchingTests.cs]: Not correct for such pattern. I see 6 matches in this code:

```php
<?php 
test_call_0();
test_call_1(a);
test_call_2(a, e);
test_call_3(c, d, a);
test_call_4(c, d, a, e);

$password = a;
call_with_password_param(1, $password, 2);

$password2 = "1234";
if ($password2->Length > 0) { }
test_call_5(1, $password2, 2);
$concat = "111" .
   "222222222"   .  
  "333";
```

Also, it's not necessary to declare a variable here.

File: Sources/PT.PM.Matching/Patterns/PatternAny.cs [hunk 2]
```diff
@@ -1,19 +1,54 @@
...
+                }
+                return context.AddMatch(ust);
             }
+            string ustString = ust.ToString();
+            var match = Regex.Match(ustString);
+            if (match.Value.Equals(ustString))
+            {
+                return context.AddMatch(ust);
+            }
+            return context.Fail();
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching/Patterns/PatternAny.cs]: This pattern can match several text spans for one UST node. Use [MatchRegex](https://github.com/PositiveTechnologies/PT.PM/blob/dev/Sources/PT.PM.Matching/PatternUtils.cs#L10) method from PatternUtils.

See example of using in: [PatternStringRegexLiteral](https://github.com/PositiveTechnologies/PT.PM/blob/dev/Sources/PT.PM.Matching/Patterns/PatternStringRegexLiteral.cs#L38).

File: Sources/PT.PM.Matching/Patterns/PatternAny.cs [hunk 3]
```diff
@@ -1,19 +1,54 @@
...
+        }
 
-            return context.AddMatch(ust);
+        public override string ToString()
+        {
+            return $@"<""{(Any ? "" : RegexString)}"">";
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching/Patterns/PatternAny.cs]: Let's use a `<#>` output for any and `<#regex#>` for regex-any.

File: Sources/PT.PM.Matching/Patterns/PatternAny.cs [hunk 4]
```diff
@@ -1,19 +1,54 @@
...
+        {
+            get => Regex.ToString();
+            set => Regex = new Regex(string.IsNullOrEmpty(value) ? Default : value, RegexOptions.Compiled);
+        }
+
+        public Regex Regex { get; private set; }
+
+        public override bool Any => Regex == null || RegexString.Equals(Default);
+
+        public PatternAny() { }
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching/Patterns/PatternAny.cs]: Please use common style: line breaks before curly bracket:

```csharp
public PatternAny()
{
}
```

File: Sources/PT.PM.Matching.Tests/PatternMatchingTests.cs (hunk 1 above)
Comment (codeFather2) [reply Sources/PT.PM.Matching.Tests/PatternMatchingTests.cs]: logic changed due to your review

File: Sources/PT.PM.Matching/Patterns/PatternAny.cs (hunk 2 above)
Comment (codeFather2) [reply Sources/PT.PM.Matching/Patterns/PatternAny.cs]: now a part of logic locate at TextUtils

File: Sources/PT.PM.Matching/Patterns/PatternAny.cs (hunk 3 above)
Comment (codeFather2) [reply Sources/PT.PM.Matching/Patterns/PatternAny.cs]: done

File: Sources/PT.PM.Matching/Patterns/PatternAny.cs (hunk 4 above)
Comment (codeFather2) [reply Sources/PT.PM.Matching/Patterns/PatternAny.cs]: sorry))

File: Sources/PT.PM.Common/TextUtils.cs [hunk 5]
```diff
@@ -63,6 +65,64 @@ public static TextSpan Union(this IEnumerable<TextSpan> textSpans)
             return resultTextSpan;
         }
 
+        public static List<TextSpan> GetCombinedTextSpan(Ust ust, TextSpan[] locations)
```
Comment (codeFather2) [inline Sources/PT.PM.Common/TextUtils.cs]: maybe it should be named `GetPreprocessedTextSpan`

File: Sources/PT.PM.Matching.Tests/PatternMatchingTests.cs [hunk 6]
```diff
@@ -110,6 +110,25 @@ public void Match_PasswordCheckInsideStatement(string patternData)
...
+                Node = new PatternAny("password")
+            };
+            patternsRepository.Add(patternsConverter.ConvertBack(new List<PatternRoot>() { pattern }));
+
+            WorkflowResult workflowResult = workflow.Process();
+            IEnumerable<MatchResultDto> matchResults = workflowResult.MatchResults.ToDto();
+            patternsRepository.Clear();
+
+            int expectedCount = 6;
+            Assert.AreEqual(expectedCount, matchResults.Count());
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching.Tests/PatternMatchingTests.cs]: Inline this variable: `Assert.AreEqual(6, matchResults.Count());`

File: Sources/PT.PM.Matching/Patterns/PatternAny.cs [hunk 7]
```diff
@@ -1,19 +1,71 @@
...
-                return context.MakeSuccess();
+                if (ust == null)
+                {
+                    return context.MakeSuccess();
+                }
+                return context.AddMatch(ust);
             }
 
-            return context.AddMatch(ust);
+            if (ust.Children.Length > 0)
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching/Patterns/PatternAny.cs]: Why did you write this line? It's totally incorrect and makes PatternAny with regex senseless.

File: Sources/PT.PM.Common/TextUtils.cs (hunk 5 above)
Comment (KvanTTT) [reply Sources/PT.PM.Common/TextUtils.cs]: `GetAlignedTextSpans`. Definitely, it should be placed to another util class, maybe `UstUtils`.
`locations` -> `matchesLocations`.

File: Sources/PT.PM.Matching/Patterns/PatternStringRegexLiteral.cs [hunk 8]
```diff
@@ -39,61 +39,8 @@ public override MatchContext Match(StringLiteral stringLiteral, MatchContext con
...
-                        }
-                    }
-
-                    if (textSpan.IsZero)
-                    {
-                        result.Add(location);
-                    }
-                }
-                matches = result;
+            {
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching/Patterns/PatternStringRegexLiteral.cs]: Excess whitespaces.

File: Sources/PT.PM.Matching/PatternRoot.cs [hunk 9]
```diff
@@ -98,7 +98,13 @@ public List<MatchResult> Match(Ust ust)
...
         private static void TraverseChildren(PatternUst patternUst, Ust ust, MatchContext context, List<MatchResult> results)
         {
-            MatchAndAddResult(patternUst, ust, context, results);
+            MatchAndAddResult(patternUst, ust, context, results, false);
+
+            if (context.Locations.Count > 0)
+            {
+                context.Locations.Clear();
+                return;
+            }
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching/PatternRoot.cs]: I think this is an incorrect addition (see below).

File: Sources/PT.PM.Matching/PatternRoot.cs [hunk 10]
```diff
@@ -98,7 +98,13 @@ public List<MatchResult> Match(Ust ust)
...
 
             if (ust != null)
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching/PatternRoot.cs]: I suggest just to add a check on `PatternAny`:

```csharp
if (ust != null && !(patternUst is PatternAny))
{
    ...
}
```

File: Sources/PT.PM.Matching/Patterns/PatternAny.cs [hunk 11]
```diff
@@ -1,19 +1,64 @@
...
+            matches = matches.Select(location => location.AddOffset(ust.TextSpan.Start)).ToArray();
+
+            return matches.Length > 0
+                ? context.AddMatches(matches)
+                : context.Fail();
+        }
+
+        public override string ToString()
+        {
+            return $@"<""#{(Any ? "" : RegexString + "#")}"">";
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching/Patterns/PatternAny.cs]: I think leading and trailing quotes can be removed: `$@"<#{(Any ? "" : RegexString + "#")}>"`.

File: Sources/PT.PM.Matching/PatternRoot.cs [hunk 12]
```diff
@@ -110,19 +116,30 @@ private static void TraverseChildren(PatternUst patternUst, Ust ust, MatchContex
...
                 var match = new MatchResult(ust.Root, context.PatternUst, context.Locations);
+
                 results.Add(match);
                 context.Logger.LogInfo(match);
             }
-            context.Locations.Clear();
+            if (clearLocations || (patternUst is PatternAnd && !context.Success))
+            {
+                context.Locations.Clear();
+            }
```
Comment (KvanTTT) [inline Sources/PT.PM.Matching/PatternRoot.cs]: I think this method can be written simpler. I'll take a look.

**Pull request merged**